### PR TITLE
[x86] cast from long w/o overflow checking

### PR DESCRIPTION
**Pull request by sejongoh**
Opened for testing

### CODE REVIEW

**Comment by mikedn** [comment]
Hmm, why do we need to do such casts in codegen? I thought that only those with overflow checks and the ones involving floating point need to be done in codegen.

**Comment by sejongoh** [comment]
Only cast_ovfl from long can reach codegen. Please ignore `genLongToIntCast` in codegen and I will open another PR later to add overflow checking to it.

**Comment by sejongoh** [comment]
This PR is just for testing and unnecessary code will be removed once testing is okay.

**Comment by sejongoh** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by sejongoh** [comment]
@dotnet-bot test OSX x64 Checked Build and Test

**Comment by sejongoh** [comment]
@BruceForstall @mikedn @adiaaida PTAL

**Comment by sejongoh** [comment]
I will update the commit description soon.

File: src/jit/decomposelongs.cpp [hunk 1]
```diff
@@ -156,6 +156,40 @@ void DecomposeLongs::DecomposeNode(GenTree** ppTree, Compiler::fgWalkData* data)
...
+                assert(genTypeSize(srcType) > genTypeSize(dstType));
+                assert(tree->gtOp.gtOp1->OperGet() == GT_LONG);
+
+                GenTree* gtLong = tree->gtOp.gtOp1;
+                GenTree* highOpd = gtLong->gtOp.gtOp2;
+                GenTree* lowOpd = gtLong->gtOp.gtOp1;
+                GenTreeStmt* curStmt = m_compiler->compCurStmt->AsStmt();
+                unsigned dstSize = genTypeSize(dstType);
+                unsigned srcSize = genTypeSize(srcType);
+
```
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: `dstSize` and `srcSize` appear to be unused.

File: src/jit/decomposelongs.cpp [hunk 2]
```diff
@@ -156,6 +156,40 @@ void DecomposeLongs::DecomposeNode(GenTree** ppTree, Compiler::fgWalkData* data)
...
+                m_compiler->fgSnipNode(curStmt, gtLong);
+                m_compiler->fgDeleteTreeFromList(curStmt, highOpd);
```
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: Is it safe to remove the entire hi tree? What if it has side-effects, a GT_ADD_HI with overflow for example?

**Comment by mikedn** [comment]
I think that a comment should be added to explain when this particular code path is hit. It seems to me that many such casts are handled during morphing by narrowing the tree. Narrowing is not always possible and then these casts reach decomposition/codegen and need to be handled somehow. 

It's quite possible that one reason why morph can fail to narrow a `long` tree is the presence of overflow checks and that would mean that `fgDeleteTreeFromList` won't work either. You may need to create an embedded statement out of the hi tree, I don't think there's any other way to have an unused tree.

File: src/jit/decomposelongs.cpp (hunk 2 above)
Comment (sejongoh) [inline src/jit/decomposelongs.cpp]: Right. How can we check side effects from `highOpd` tree? Can it be simply done by checking the side effect flag on 'highOpd` node?

**Comment by sejongoh** [comment]
Yeah.. morph seems to narrow the tree in many cases. I couldn't get a chance to look at why morph fails to narrow some cases. In one case that I hit, there is no overflow check. If the transformation in morph is just an optimization, I think that we need to keep this anyway.

```
*  stmtExpr  void  (top level) (IL 0x000...  ???)
|     /--*  lclVar    int    V00 arg0         u:2 (last 
|     |        /--*  lclVar    int    V01 arg1         u:2
|     |     /--*  cast      int <- ubyte <- int $141
|     |  /--*  cast      long <- ulong <- uint $180
|     +--*  cast      int <- uint <- long $142
|  /--*  |         int    $143
\--*  st.lclVar int    V02 tmp0         d:2
```

File: src/jit/decomposelongs.cpp (hunk 1 above)
Comment (sejongoh) [inline src/jit/decomposelongs.cpp]: Thanks for finding these. I forgot to remove them..

File: src/jit/decomposelongs.cpp (hunk 2 above)
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: >  Can it be simply done by checking the side effect flag on 'highOpd` node?

Yes, side-effect flags apply to the entire tree, not just the node. 

Assuming that the side-effect flags are correct in the first place, decomposition doesn't always get this right currently...

**Comment by mikedn** [comment]
> I think that we need to keep this anyway

Yes, it just needs to deal correctly with side-effects. Otherwise it seems fine to me.

**Comment by sejongoh** [comment]
To create an embedded statement, does it have to some store node such as st.lclVar or storeInd? Or, can any node be the root node? Another option is to let codegen handle this, and the decomposition just optimizes cases without side-effects.

**Comment by mikedn** [comment]
> To create an embedded statement, does it have to some store node such as st.lclVar or storeInd? Or, can any node be the root node?

You should not need a store, the JIT can handle nodes that produce a value that's not used and this kind of thing already happens in other situations. For example if you have `int x = checked(a - b);` and `x` isn't used anywhere then the store will be eliminated and the GT_SUB will stay because of its side effect, the result of GT_SUB is unused.

> Another option is to let codegen handle this, and the decomposition just optimizes cases without side-effects.

IMO it's not worth complicating codegen for this. And if we move to LIR then all the complications caused by embedded statements will go away anyway.

**Comment by sejongoh** [comment]
It doesn't complicate codegen. It will anyway have logic to generate code for casts from long to int to handle cast_ovfl. Codegen can simply skip generating instructions for overflow checking. Once we implement codegen for cast_ovfl from long to int, the decomposition changes in this PR will be just an optimization.

File: src/jit/codegenxarch.cpp [hunk 3]
```diff
@@ -2056,12 +2056,17 @@ CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
 
     case GT_CAST:
 #if !defined(_TARGET_64BIT_)
-        // We will NYI in DecomposeNode() if we are cast TO a long type, but we do not
-        // yet support casting FROM a long type either, and that's simpler to catch
+        // We will NYI in DecomposeNode() if we are cast TO a long type. 
+        // Casts from long to int/short/byte without overflow check are handled 
+        // in DecomposeNode(). We do not yet support other casts from long.
```
Comment (adiaaida) [inline src/jit/codegenxarch.cpp]: Should the period at the end of this line be a comma?

File: src/jit/decomposelongs.cpp [hunk 4]
```diff
@@ -156,6 +156,51 @@ void DecomposeLongs::DecomposeNode(GenTree** ppTree, Compiler::fgWalkData* data)
 
     if (tree->TypeGet() != TYP_LONG)
     {
+        // For long -> int/short/byte conversions, if overflow checking is not needed,
+        // the upper 32 bits can be simply dropped. If it is not dropped here, codegen will 
+        // generate unnecessary instructions to compute the upper 32 bits.
+        // In many cases, morph optimizes this kind of casts but some cases slip to here.
```
Comment (adiaaida) [inline src/jit/decomposelongs.cpp]: either casts -> cast or this->these

File: src/jit/decomposelongs.cpp [hunk 5]
```diff
@@ -156,6 +156,51 @@ void DecomposeLongs::DecomposeNode(GenTree** ppTree, Compiler::fgWalkData* data)
...
+        // An example tree reaching here is:
+        // *  stmtExpr  void  (top level) (IL 0x000... ? ? ? )
+        // | / --*  lclVar    int    V00 arg0         u : 2 (last
+        // | | / --*  lclVar    int    V01 arg1         u : 2
+        // | | / --*  cast      int <-ubyte <-int $141
+        // | | / --*  cast      long <-ulong <-uint $180
+        // | +--*  cast      int <-uint <-long $142
+        // | / --* | int    $143
+        // \--*  st.lclVar int    V02 tmp0         d : 2
+        if (tree->OperGet() == GT_CAST && !tree->gtOverflow())
```
Comment (adiaaida) [inline src/jit/decomposelongs.cpp]: Could we put this in its own function?

**Comment by RussKeldorph** [comment]
@pgavlin @BruceForstall Is there anything useful in here after the LIR changes, or should we close this out and start over?

**Comment by BruceForstall** [comment]
Maybe. But probably should just start over.

**Comment by pgavlin** [comment]
I agree; starting over is probably the right call.

It would be nice to enumerate which sorts of casts are not yet implemented if we haven't already done so. @mikedn also has a PR out for some casts here: https://github.com/dotnet/coreclr/pull/4986

**Pull request closed without merging**